File: BillPayments/Features/Billers/Data/BillerCatalogStore.swift
import Foundation
import Observation

@Observable
final class BillerCatalogStore {
    private(set) var catalog: Async<BillerCatalog> = .loading(previous: nil)
    @ObservationIgnored private(set) var pendingWork: Task<Void, Never>?
    @ObservationIgnored private(set) var observationWork: Task<Void, Never>?

    @ObservationIgnored private let repository: BillerRepository
    @ObservationIgnored private let settingsStore: SettingsStore
    @ObservationIgnored private let freshness: Freshness

    init(repository: BillerRepository, settingsStore: SettingsStore, clock: Clock) {
        self.repository = repository
        self.settingsStore = settingsStore
        self.freshness = Freshness(clock: clock, maxAge: 30 * 60)
        observationWork = Task { [weak self, settingsStore] in
            for await language in Observations({ settingsStore.language }) {
                guard let self else { return }
                self.fetch(language: language)
            }
        }
    }

    private func fetch(language: String) {
        pendingWork?.cancel()
        catalog = .loading(previous: catalog.valueOrNull)
        pendingWork = Task { [weak self] in
            guard let self else { return }
            do {
                let result = try await self.repository.fetchCatalog(language: language)
                if Task.isCancelled { return }
                self.catalog = .data(result)
                self.freshness.markFetched()
            } catch {
                if Task.isCancelled { return }
                self.catalog = .error(error)
            }
        }
    }

    func refreshIfStale() {
        if case .loading = catalog { return }
        if !freshness.isStale() { return }
        fetch(language: settingsStore.language)
    }

    func retry() {
        fetch(language: settingsStore.language)
    }
}

File: app/src/main/java/com/billpayments/features/billers/data/BillerCatalogStore.kt
package com.billpayments.features.billers.data

import com.billpayments.core.async.Async
import com.billpayments.core.cache.Freshness
import com.billpayments.core.time.Clock
import com.billpayments.features.settings.data.SettingsStore
import java.time.Duration
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch

@Singleton
class BillerCatalogStore @Inject constructor(
    private val repository: BillerRepository,
    private val settingsStore: SettingsStore,
    clock: Clock,
    private val scope: CoroutineScope,
) {
    private val freshness = Freshness(clock, Duration.ofMinutes(30))
    private val _catalog = MutableStateFlow<Async<BillerCatalog>>(Async.Loading)
    val catalog: StateFlow<Async<BillerCatalog>> = _catalog.asStateFlow()

    init {
        scope.launch {
            settingsStore.language.collectLatest { fetch(it) }
        }
    }

    private suspend fun fetch(language: String) {
        _catalog.value = Async.Loading
        try {
            _catalog.value = Async.Data(repository.fetchCatalog(language))
            freshness.markFetched()
        } catch (e: Exception) {
            _catalog.value = Async.Error(e)
        }
    }

    fun refreshIfStale() {
        if (_catalog.value is Async.Loading || !freshness.isStale()) return
        scope.launch { fetch(settingsStore.language.value) }
    }

    fun retry() {
        scope.launch { fetch(settingsStore.language.value) }
    }
}
